lib/msgServer/mmrp/index.js

```diff
@@ -130,28 +147,64 @@ exports.MmrpNode = MmrpNode;
  * @param {Function} [cb]
  */
 
-MmrpNode.prototype.connect = function (uri, clusterId, cb) {
+MmrpNode.prototype.connect = function (uri, data, cb) {
+	const clusterId = data.clusterId;
+	const timestamp = data.timestamp;
+
 	// we cannot connect through routers (long routes), we are establishing a direct connection
 
 	assert.strictEqual(typeof uri, 'string');
 	assert.strictEqual(typeof clusterId, 'string');
 
-	var route = [clusterId];
-	var relay = this.relays[clusterId];
+	// check if other relay(s) with the same URI exist, and disconnect them;
+	// this is required to clean up old relays that have yet to be removed
+	// from the relay list
+
+	for (const relayClusterId in this.relays) {
+		const relay = this.relays[relayClusterId];
+
+		// Filter out our own relay
+		if (relayClusterId === clusterId) {
+			continue;
+		}
+
+		// Filter out relays with a different URI
+		if (relay.uri !== uri) {
+			continue;
+		}
+
+		// If we find a more recent node in our
+		// list, we should then ignore adding
+		// the node we have just received overall
+		if (relay.timestamp >= timestamp) {
+			return callMeMaybe(cb);
+		}
+
+		logger.verbose.data({
+			oldRelayId: relayClusterId,
+			newRelayId: clusterId
+		}).log(this.clusterId, 'found relay with same uri, removing the relay');
+
+		// We delete the relay reference, but do not disconnect; quick disconnect-reconnect
+		// appear to confuse ZeroMQ, and may permanently break the connection in some cases.
```

Comment:
Some more information, perhaps documentation or a link to a gist, or .. you name it, would be really helpful for future generations.